src/xkernel.cpp

```diff
@@ -61,29 +64,64 @@ namespace xeus
                      interpreter_ptr interpreter,
                      history_manager_ptr history_manager,
                      server_builder builder)
-        : m_config(config)
-        , m_user_name(user_name)
-        , p_interpreter(std::move(interpreter))
-        , p_history_manager(std::move(history_manager))
-        , m_builder(builder)
+        : m_config(config),
+          m_user_name(user_name),
+          p_interpreter(std::move(interpreter)),
+          p_history_manager(std::move(history_manager)),
+          m_builder(builder),
+          m_empty_config(false)
+    {
+        init();
+    }
+
+    xkernel::xkernel(const std::string& user_name,
+                     interpreter_ptr interpreter,
+                     history_manager_ptr history_manager,
+                     server_builder builder)
+        : m_user_name(user_name),
+          p_interpreter(std::move(interpreter)),
+          p_history_manager(std::move(history_manager)),
+          m_builder(builder),
+          m_empty_config(true)
     {
+        init();
     }
 
     xkernel::~xkernel() {}
 
-    void xkernel::start()
+    void xkernel::init()
     {
         m_kernel_id = new_xguid();
         m_session_id = new_xguid();
 
+        if (m_config.m_key.size() == 0)
+        {
+            std::string ascii = "0123456789abcdefghijklmnopqrstuvwxyz";
+
+            std::random_device r;
+            std::default_random_engine generator(r());
+            std::uniform_int_distribution<std::size_t> distribution(0, 35);
+
+            m_config.m_key.resize(33);
+            for (auto& ch: m_config.m_key)
+            {
+                ch = ascii[distribution(generator)];
+            }
+            m_config.m_key[8] = '-';
+        }
+
         using authentication_ptr = xkernel_core::authentication_ptr;
         authentication_ptr auth = make_xauthentication(m_config.m_signature_scheme, m_config.m_key);
 
-        zmq::multipart_t start_msg;
-        build_start_msg(auth, m_kernel_id, m_user_name, m_session_id, start_msg);
+        build_start_msg(auth, m_kernel_id, m_user_name, m_session_id, m_start_msg);
 
         p_server = m_builder(m_context, m_config);
 
+        if (m_empty_config)
```

Comment:
I think this is a bit of overkill, testing if one of the port of `m_config` should be enough and would avoid adding a data member used for initialization only.

Reply:
Yeah.. I also thought about testing one of the port, but I don't feel like it's a clean solution. I think the cleanest way to do it would be to add a `config_needs_update` flag in xserver, and if I set one port randomly I set it to true

Reply:
Indeed!

Reply:
But, the problem is that this logic should be in `xserver_zmq` and `xserver`

Reply:
Why is it a problem? You have two options here:
- either you consider this logic should be in the `xserver` hierarchy. Then `xserver` provides a `needs_config_update` API method, which needs to be implemented in `xserver_zmq` (so the flag should be in `xsever_zmq` and not `xserver` actually).
- or you consider that the configuration needs to be updated as soon as one of its port is empty and you implement this logic in `xkernel`. I agree that testing only one port is a bit of a hack, but testing them all is an acceptable solution for me.

Reply:
I went for the first option :)